packages/adblocker/src/resources.ts

```diff
@@ -25,9 +25,101 @@ export interface Resource {
   body: string;
 }
 
-// TODO - support # alias
+type ResourceBase = {
+  names: string[];
+  content: string;
+};
+
+export type RedirectResource = ResourceBase & {
+  contentType: string;
+  encoding?: 'base64';
+};
+
+export type Scriptlet = ResourceBase & {
+  dependencies: string[];
+  executionWorld: 'main' | 'isolated';
+  requiresTrust: boolean;
+};
+
+export type ResourcesDistribution = {
+  redirects: RedirectResource[];
+  scriptlets: Scriptlet[];
+};
+
 // TODO - support empty resource body
 
+export const SCRIPTLET_BODY_PREFIX = `if (typeof scriptletGlobals === 'undefined') {
+  var scriptletGlobals = {};
+}
+`;
+
+export const SCRIPTLET_BODY_SUFFIX = `(...['{{1}}','{{2}}','{{3}}','{{4}}','{{5}}','{{6}}','{{7}}','{{8}}','{{9}}','{{10}}'].filter((a,i) => a !== '{{'+(i+1)+'}}').map((a) => decodeURIComponent(a)));`;
+
+function assembleScriptlet(scriptlet: Scriptlet, scriptlets: Map<string, Scriptlet>) {
+  const stack: Set<Scriptlet> = new Set();
+  let queue: Scriptlet[] = [scriptlet];
+
+  while (queue.length > 0) {
+    const current = queue.pop()!;
+    stack.add(current);
+
+    for (const dependencyName of current.dependencies) {
+      const dependency = scriptlets.get(dependencyName);
+      if (dependency === undefined) {
+        return `throw new Error('Failed to resolve scriptlet dependency: Unable to locate: ${current.names[0]} -> ${dependencyName}');`;
+      }
+
+      if (stack.has(dependency) === false) {
+        queue.push(dependency);
+        stack.add(dependency);
+      }
+    }
+  }
+
+  queue = Array.from(stack);
+
+  let body = SCRIPTLET_BODY_PREFIX;
+  let functionName = '';
+
+  while (queue.length !== 0) {
+    const entry = queue.pop()!;
```

Comment:
Updated. The order of dependencies are now reversed in the result but I don't think it'll be a problem since JavaScript doesn't complain about the functions' order.